### PR TITLE
eth, les: rework chain tracer

### PR DESCRIPTION
**Pull request by rjl493456442**
This PR simplifies the logic of chain tracer and also adds the unit tests.

The most important change has been made in this PR is the state management. Whenever a tracing state is acquired there is a corresponding release function be returned as well. It must be called once the state is used up, otherwise resource leaking can happen.

And also the logic of state management has been simplified a lot. Specifically, the state provider(eth backend, les backend) should ensure the state is available and referenced. State customers can use the state according to their own needs, or build other states based on the given state. But once the release function is called, there is no guarantee of the availability of the state.

### CODE REVIEW

File: eth/tracers/api.go [hunk 1]
```diff
@@ -53,13 +53,6 @@ const (
 	// and reexecute to produce missing historical state necessary to run a specific
 	// trace.
 	defaultTraceReexec = uint64(128)
-
-	// defaultTracechainMemLimit is the size of the triedb, at which traceChain
-	// switches over and tries to use a disk-backed database instead of building
-	// on top of memory.
-	// For non-archive nodes, this limit _will_ be overblown, as disk-backed tries
-	// will only be found every ~15K blocks or so.
-	defaultTracechainMemLimit = common.StorageSize(500 * 1024 * 1024)
```
Comment (holiman) [inline eth/tracers/api.go:62]: I see that this is now removed. Some context for this thing -- if we do chain tracing from an early block, and want to e.g. trace 1M blocks between 12M and 13M on an archive node. 
Then the way it works is that it will dig up state for 12M, and trace with that. After that, it will continue operating on the same state (now with modifications from that one block). It will continue, using in-memory trie database as state-backend, until it hits the `defaultTracechainMemLimit`. At this point, e.g. `120001000`, it will release that state, and load a new one from disk. 

This means that it is both pretty fast (reuses the in-memory state instead of always loading from disk) and also does not go out of memory (since it doesn't accumulate indefinitely). I implemented this in https://github.com/ethereum/go-ethereum/pull/23736 . 

So, I'm curious, how is this scenario handled in this PR? "Always reload", "never reload" or something else?

Comment (rjl493456442) [reply eth/tracers/api.go:62]: Now we will always try to find the state in disk first. The logic is here https://github.com/rjl493456442/go-ethereum/blob/improve-chain-tracer/eth/state_accessor.go#L81

I totally agree that we should reuse state in disk as much as possible. So I make it as the default behavior.

Comment (holiman) [reply eth/tracers/api.go:62]: > I totally agree that we should reuse state in disk as much as possible. So I make it as the default behavior.

...I'm not 100% certain that we're in agreement.. (?)
So what I'm saying is the opposite: we want to avoid starting 'fresh' from the disk-state most of the time. Only do it when memory constraints force us to. 


--- 
Now, I might be ok with doing a cleaner-but-slower approach if it makes the code and the semantics about "state loading" simpler, but if we go down that route, we should definitely measure it first, to see how much slower it becomes.

Comment (rjl493456442) [reply eth/tracers/api.go:62]: Get it.

For archive node, I think in-memory solution should have a lot of benefits. `cleaner-but-slower` approach doesn't make too much sense. Thanks for pointing it out, I will think about it.

Comment (rjl493456442) [reply eth/tracers/api.go]: I have added it back.

But the behavior is slightly different. Previously whenever the cached nodes reach the threshold, we switch over to `preferDisk` mode. It can be inefficient in archive node which always build the fresh trie database afterwards. Now this mode will only be enabled once the memory usage exceeds the limit.

Please take another look.

**Comment by rjl493456442** [comment]
I am syncing a goerli node at the moment and want to manually confirm chain tracing is not broken.

**Comment by LarrySOScary** [comment]
I think this may have to do with me, can somebody guide me how to upgrade my node properly?

**Comment by rjl493456442** [comment]
@holiman I have carefully compared the memory usage/trace output with master. I am confident about this PR now.

- Memory usage

Both master and PR will use additional 2,3GB memory after 40 minutes tracing, but the the real held memory(dirty cache of triedb) is limited by 500MB properly. So I guess it's just because the Golang GC doesn't release the memory very soon.

- Master is killed

In my test scenario(goerli archive node), when the dirty cache size exceeds 500MB, it fallbacks to disk state. In this case master will **always** use disk state without proper releasing, eventually it's killed because of high memory usage.

PR doesn't have this issue, can trace 70K blocks successfully.


TL;DR, I think the PR itself is good, please take another look.

**Comment by tuanggolt** [comment]
open

File: eth/state_accessor.go [hunk 2]
```diff
@@ -30,73 +30,90 @@ import (
...
 // StateAtBlock retrieves the state database associated with a certain block.
 // If no state is locally available for the given block, a number of blocks
 // are attempted to be reexecuted to generate the desired state. The optional
-// base layer statedb can be passed then it's regarded as the statedb of the
+// base layer statedb can be provided which is regarded as the statedb of the
 // parent block.
+//
+// An additional release function will be returned if the requested state is
+// available. The release is expected be invoked when the state is used, though
+// it can be noop in some cases, otherwise the resources leaking can happen.
```
Comment (s1na) [inline eth/state_accessor.go:45]:
```suggestion
// An additional release function will be returned if the requested state is
// available. Release is expected to be invoked when the returned state is no longer needed.
// Its purpose is to prevent resource leaking. Though it can be noop in some cases.
```

File: eth/state_accessor.go [hunk 3]
```diff
@@ -155,28 +178,28 @@ func (eth *Ethereum) StateAtBlock(block *types.Block, reexec uint64, base *state
 		nodes, imgs := database.TrieDB().Size()
 		log.Info("Historical state regenerated", "block", current.NumberU64(), "elapsed", time.Since(start), "nodes", nodes, "preimages", imgs)
 	}
-	return statedb, nil
+	return statedb, func() { database.TrieDB().Dereference(block.Root()) }, nil
```
Comment (karalabe) [inline eth/state_accessor.go:181]: I'm unsure about this. You're returning a deref call, but there's no matching ref call added. Won't this end up us dereferencing something live?

Comment (rjl493456442) [reply eth/state_accessor.go:181]: The reference work is done in the for loop.

Once we enter the loop for executing blocks, we will always reference the current one and reference the parent. And we can ensure *at least one block* is executed. In the end, we will keep the reference of the latest one.

Comment (karalabe) [reply eth/state_accessor.go:181]: Ah ok, I didn't get yet to the part of TraceChain where you removed the dereference and replaced it with this :+1:

File: eth/tracers/api.go [hunk 4]
```diff
@@ -329,105 +365,97 @@ func (api *API) traceChain(ctx context.Context, start, end *types.Block, config
...
-				s1, s2 := trieDb.Size()
-				if !preferDisk && (s1+s2) > defaultTracechainMemLimit {
-					log.Info("Switching to prefer-disk mode for tracing", "size", s1+s2)
-					preferDisk = true
-				}
-			}
-			parent = block.Root()
-
-			next, err := api.blockByNumber(localctx, rpc.BlockNumber(number+1))
+			next, err := api.blockByNumber(ctx, rpc.BlockNumber(number+1))
```
Comment (karalabe) [inline eth/tracers/api.go:404]: If this errors, shouldn't we call `release()`? If was just returned above and we're only adding it to the task way below / calling on closure. But if we hit this error, it'd be a resource leak.

Comment (rjl493456442) [reply eth/tracers/api.go]: thanks for catching, fixed.
